pkg/backends/cloudwatch/cloudwatch.go

```diff
@@ -170,58 +198,128 @@ func (client Client) buildMetricData(metrics *gostatsd.MetricMap) (metricData []
 	return metricData
 }
 
-// SendMetricsAsync sends the metrics in a MetricsMap to AWS Cloudwatch,
-// preparing payload synchronously but doing the send asynchronously.
-func (client Client) SendMetricsAsync(ctx context.Context, metrics *gostatsd.MetricMap, cb gostatsd.SendCallback) {
-	api := client.cloudwatch
-	metricData := client.buildMetricData(metrics)
-	length := len(metricData)
-	errors := []error{}
+// setupFlushTimer will either increase the timeout on the flush timer, or create the flush timer if it doesn't
+// exist yet.  It returns true or false to indicate whether or not the caller is responsible for servicing the
+// timer when it actually expires (true if it should).  Thread safe.
+func (client *Client) setupFlushTimer() bool {
+	client.timerMutex.Lock()
+	defer client.timerMutex.Unlock()
+	if client.timer != nil {
+		client.timer.Reset(client.batchDuration)
+		return false
+	} else {
+		client.timer = time.NewTimer(client.batchDuration)
+		return true
+	}
+}
 
-	if length < 1 {
-		cb(errors)
+// clearFlushTimer will stop the flush timer.  Thread safe.
+func (client *Client) clearFlushTimer() {
+	client.timerMutex.Lock()
+	client.timer = nil // Stop current timer
+	client.timerMutex.Unlock()
+}
+
+func (client *Client) flush() {
+	if !client.setupFlushTimer() {
 		return
 	}
 
-	go func() {
-		start := 0
+	// This is safe to read outside the timerMutex, as only the
+	// goroutine that created the initial timer can ever be here
+	<-client.timer.C
+
+	// After this another goroutine could potentially run, start a
+	// timer, have that timer expire, and collect the data we were
+	// going to send - but that's ok.
+	client.clearFlushTimer()
+
+	// Copy pending metric items and create new queue instance
+	// client.queue is protected by a Read-Write Lock such
+	// that execution is only blocked while replacing the queue
+	client.queueMutex.Lock()
+	queue := client.queue
+	client.queue = []*QueueItem{}
+	client.queueMutex.Unlock()
+
+	// Run actual function
+	metricData := []*cloudwatch.MetricDatum{}
```

Comment:
Good idea

- [ ] preallocate queue

Reply:
It also may or may not be beneficial to have a slice of values rather than pointers. Of course if they are already pointers when they reach this code then it may not make any difference.